### PR TITLE
Enhance/slm streaming update

### PR DESCRIPTION
**Pull request by Killthebug**
This PR:
1. Updates `chat_completions()` for SLMs to return a co-routine instead of a synchronous wrapper. This change was made to be inline with the new method of e2e inference as defined by : #78 
2. Fixed minor bug in encoder

### CODE REVIEW

File: prem_utils/connectors/base.py [hunk 1]
```diff
@@ -52,7 +52,7 @@ def get_number_of_tokens_request(self, messages: list):
 
     def get_number_of_tokens_response(self, text: str):
         encoding = tiktoken.get_encoding("cl100k_base")
-        return len(encoding.encode(text))
+        return len(encoding.encode(text, disallowed_special=()))
```
Comment (nsosio) [inline prem_utils/connectors/base.py:55]: what's this for?

File: prem_utils/connectors/prem.py [hunk 2]
```diff
@@ -44,98 +41,78 @@ def parse_chunk(self, chunk):
...
-                        # Sometimes (randomly) chunks contains multiple lines
-                        lines = chunk.strip().split("\n")
-                        for line in lines:
-                            chunk_queue.put(json.loads(line))
-        chunk_queue.put(None)  # Signal completion
+                    buffer += chunk
+                    while "\n" in buffer:
+                        line, buffer = buffer.split("\n", 1)
+                        if line.strip():
+                            yield json.loads(line)
```
Comment (nsosio) [inline prem_utils/connectors/prem.py:94]: why do we need a buffer?

File: prem_utils/connectors/base.py (hunk 1 above)
Comment (Killthebug) [reply prem_utils/connectors/base.py:55]: If a text being tokenized by tiktoken contains a special token like `<|endoftext|>`, we will see the error:
```
ValueError: Encountered text corresponding to disallowed special token '<|endoftext|>'.
If you want this text to be encoded as a special token, pass it to `allowed_special`, e.g. `allowed_special={'<|endoftext|>', ...}`.
If you want this text to be encoded as normal text, disable the check for this token by passing `disallowed_special=(enc.special_tokens_set - {'<|endoftext|>'})`.
To disable this check for all special tokens, pass `disallowed_special=()`.
```

I observed that one of the SLMs generated the `<|endoftext|>` token. Hence to deal with it, I added this. It doesn't impact other tokens.

File: prem_utils/connectors/prem.py (hunk 2 above)
Comment (Killthebug) [reply prem_utils/connectors/prem.py:94]: Yea, not very intuitive, but in my experiments, I observed that sometimes the response from prem-ml yielded incomplete chunks. Due to this, the json parsing of the line would error out. To deal with this, I added a buffer here to cache in incomplete responses and put them together. 

This was an edge case that was observed only a few times during experimenting. 

I also checked the streaming function in prem-ml  ([link](https://github.com/premAI-io/prem-ml/blob/406105d5668ff04dcb6ff67625328f7ad97f3109/prem_ml/core/services.py#L150C4-L161C86)) and it yields only a token at a time. Couldn't understand why this was happening, so I decided to add a buffer in.

-------

`prem-ml stream_generator()`
```
    def stream_generator():
        try:
            for response in modal_model.completion.remote_gen(data, stream=True):
                yield json.dumps(prepare_chat_response_chunk(response=response["content"], model=model)) + "\n"
        except httpx.RequestError as error:
            yield json.dumps(
                {"status": error.response.status_code if error.response else None, "error": str(error)}
            ) + "\n"
        except ValueError as error:
            yield json.dumps({"status": 500, "error": str(error)}) + "\n"

    return StreamingHttpResponse(stream_generator(), content_type="application/json")
```

**Review by nsosio — APPROVED** [review]

**Pull request merged**